enclave/src/lib.rs

```diff
@@ -522,136 +550,186 @@ where
+
+		debug!("Got following trusted calls from pool: {:?}", trusted_calls);
+		// call execution
+		for trusted_call_signed in trusted_calls.into_iter() {
+			match handle_trusted_worker_call(
+				&mut calls,
+				&mut state,
+				&trusted_call_signed,
 				latest_onchain_header.clone(),
-				call_hashes,
 				shard,
-				prev_state_hash,
-				&mut state,
+				on_chain_ocall,
 			) {
-				Ok((block_confirm, signed_block)) => {
-					calls.push(block_confirm);
-					blocks.push(signed_block.clone());
-
-					// Notify watching clients of InSidechainBlock
-					let composed_block = signed_block.block();
-					let block_hash: BlockHash = blake2_256(&composed_block.encode()).into();
-					pool.pool()
-						.validated_pool()
-						.on_block_created(composed_block.signed_top_hashes(), block_hash);
+				Ok(hashes) => {
+					let inblock = match hashes {
+						Some((_, operation_hash)) => {
+							call_hashes.push(operation_hash);
+							true
+						},
+						None => {
+							// remove call as invalid from pool
+							false
+						},
+					};
+
+					// TODO: prune instead of remove_top ? Block needs to be known
+					// TODO: move this pruning to after finalization confirmations, not here!
+					// remove calls from pool (either as valid or invalid)
+					author
+						.remove_top(
+							vec![TrustedOperationOrHash::Operation(
+								trusted_call_signed.into_trusted_operation(true),
+							)],
+							shard,
+							inblock,
+						)
+						.unwrap();
 				},
-				Err(e) => error!("Could not compose block confirmation: {:?}", e),
+				Err(e) =>
+					error!("Error performing worker call (will not push top hash): Error: {:?}", e),
+			};
+			// Check time
+			if time_is_overdue(Timeout::Call, start_time) {
+				is_done = true;
+				break
 			}
-			// save updated state after call executions
-			let _new_state_hash = state::write(state.clone(), &shard)?;
+		}
+		// create new block (side-chain)
+		match compose_block_and_confirmation(
+			latest_onchain_header.clone(),
+			call_hashes,
+			shard,
+			prev_state_hash,
+			&mut state,
+		) {
+			Ok((block_confirm, signed_block)) => {
+				calls.push(block_confirm);
+				blocks.push(signed_block.clone());
+
+				// Notify watching clients of InSidechainBlock
+				let composed_block = signed_block.block();
+				let block_hash: BlockHash = blake2_256(&composed_block.encode()).into();
+				pool.pool()
+					.validated_pool()
+					.on_block_created(composed_block.signed_top_hashes(), block_hash);
+			},
+			Err(e) => error!("Could not compose block confirmation: {:?}", e),
+		}
+		// save updated state after call executions
+		let _new_state_hash = state::write(state.clone(), &shard)?;
+
+		if is_done {
+			break
+		}
+	}
+
+	Ok((calls, blocks))
+}
 
-			if is_done {
+fn load_initialized_state(shard: &H256) -> SgxResult<State> {
+	let state = if state::exists(&shard) {
+		state::load(&shard)?
+	} else {
+		state::init_shard(&shard)?;
+		Stf::init_state()
+	};
+	Ok(state)
+}
```

Comment:
This was duplicated code (3 times, reported by CLion) - extracted into a function that is shared.